bob-access/src/authenticator/basic.rs

```diff
@@ -1,62 +1,141 @@
-                        cred.kind().map(|k| k.is_internode()) == Some(true)))
-        {
-            self.nodes = nodes;
+        if Self::node_creds_ok(&nodes) {
+            let mut resolved = HashMap::new();
+            for (nodename, cred) in nodes.drain() {
+                if cred.ip().is_none() && cred.hostname().is_some() {
+                    self.spawn_resolver(cred);
+                } else {
+                    resolved.insert(nodename, cred);
+                }
+            }
+            let mut nodes_creds = self.nodes.write().expect("nodes credentials lock");
+            nodes_creds.extend(resolved.drain());
             Ok(())
         } else {
             let message = "nodes credentials missing ip or node name";
             Err(Error::CredentialsNotProvided(message.to_string()))
         }
     }
 
-    fn check_node_request(&self, node_name: &String, ip: Option<IpAddr>) -> Option<bool> {
-        if self.nodes.is_empty() {
-            warn!("nodes credentials not set");
+    fn spawn_resolver(&self, cred: Credentials) {
+        tokio::spawn(Self::resolve_worker(self.nodes.clone(), cred, self.resolve_sleep_dur_ms));
+    }
+
+    async fn resolve_worker(nodes: Arc<RwLock<NodesCredentials>>, mut cred: Credentials, sleep_dur_ms: u64) {
+        let hostname = cred.hostname().as_ref().expect("resolve worker without hostname");
+        let mut addr: Option<Vec<SocketAddr>> = None;
+        for _ in 0..60 {
+            addr = match lookup_host(hostname).await {
+                Ok(address) => Some(address.collect()),
+                _ => None
+            };
+            if let Some(addr) = addr.as_ref() {
+                if addr.len() > 0 {
+                    break;
+                }
+            }
+
+            tokio::time::sleep(Duration::from_secs(1)).await;
+        }
+
+        while addr.is_none() || (addr.is_some() && addr.as_ref().unwrap().len() == 0) {
+            tokio::time::sleep(Duration::from_millis(sleep_dur_ms)).await;
+
+            addr = match lookup_host(hostname).await {
+                Ok(address) => Some(address.collect()),
+                _ => None
+            };
+        }
+
+        let addr = addr.expect("somehow addr is none");
+        cred.set_addresses(addr);
+        let mut nodes = nodes.write().expect("nodes credentials lock");
+        if let Some(CredentialsKind::InterNode(nodename)) = cred.kind() {
+            nodes.insert(nodename.clone(), cred);
+        } else {
+            error!("resolved credentials are not internode");
         }
-        self.nodes
-            .get(ip.as_ref()?)
-            .map(|creds| {
-                for cred in creds {
-                    if let Some(CredentialsKind::InterNode(other_name)) = cred.kind() {
-                        if node_name == other_name {
+    }
+
+    fn mark_unresolved(&self, nodename: &str) {
+        let mut nodes = self.nodes.write().expect("nodes credentials lock");
+        let cred = nodes.remove(nodename);
```

Comment:
From the common logic perspective it is safer not to remove the node from the list. It is better just to update the value and mark it as having unresolved hostname

Reply:
Also, deletion is not recommended because requests may come from a fake node with the wrong address. If we delete immediately, then other requests from the real node will stop coming. This opens the door for DDoS

Reply:
Also, to prevent DDoS it would be good to start re-resolving not on the first request with unknown ip, but on the sequence of e.g. 3 such requests (without any requests from correct ip)